src/bin/pg_autoctl/nodestate_utils.c

```diff
@@ -0,0 +1,306 @@
+{
+	(void) prepareHostNameSeparator(headers->nameSeparatorHeader,
+									headers->maxNameSize);
+
+	(void) prepareHostNameSeparator(headers->hostSeparatorHeader,
+									headers->maxHostSize);
+
+	(void) prepareHostNameSeparator(headers->nodeSeparatorHeader,
+									headers->maxNodeSize);
+}
+
+
+/*
+ * re-compute headers properties from current properties and the new node
+ * characteristics.
+ */
+void
+nodestateAdjustHeaders(NodeAddressHeaders *headers,
+					   NodeAddress *node, int groupId)
+{
+	int nameLen = strlen(node->name);
+
+	/* compute strlen of host:port */
+	IntString portString = intToString(node->port);
+	int hostLen =
+		strlen(node->host) + strlen(portString.strValue) + 1;
+
+	/* compute strlen of groupId/nodeId, as in "0/1" */
+	IntString nodeIdString = intToString(node->nodeId);
+	int nodeLen = 0;
+
+	switch (headers->nodeKind)
+	{
+		case NODE_KIND_STANDALONE:
+		{
+			nodeLen = strlen(nodeIdString.strValue);
+			break;
+		}
+
+		default:
+		{
+			IntString groupIdString = intToString(groupId);
+
+			nodeLen =
+				strlen(groupIdString.strValue) +
+				strlen(nodeIdString.strValue) + 1;
+
+			break;
+		}
+	}
+
+	/* initialize to mininum values, if needed */
+	if (headers->maxNameSize == 0)
+	{
+		/* Name */
+		headers->maxNameSize = 5;
+	}
+
+	if (headers->maxHostSize == 0)
+	{
+		/* Host:Port */
+		headers->maxHostSize = 10;
+	}
+
+	if (headers->maxNodeSize == 0)
+	{
+		/* groupId/nodeId */
+		headers->maxNodeSize = 5;
+	}
+
+	if (nameLen > headers->maxNameSize)
+	{
+		headers->maxNameSize = nameLen;
+	}
+
+	if (hostLen > headers->maxHostSize)
+	{
+		headers->maxHostSize = hostLen;
+	}
+
+	if (nodeLen > headers->maxNodeSize)
+	{
+		headers->maxNodeSize = nodeLen;
+	}
+}
+
+
+/*
+ * nodestatePrintHeader prints the given CurrentNodeStateArray header.
+ */
+void
+nodestatePrintHeader(NodeAddressHeaders *headers)
+{
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
+			headers->maxNameSize, "Name",
+			headers->maxNodeSize, "Node",
+			headers->maxHostSize, "Host:Port",
+			"Current State", "Assigned State",
+			"LSN", "Health");
+
+	fformat(stdout, "%*s-+-%*s-+-%*s-+-%17s-+-%17s-+-%17s-+-%6s\n",
+			headers->maxNameSize,
+			headers->nameSeparatorHeader,
+			headers->maxNodeSize,
+			headers->nodeSeparatorHeader,
+			headers->maxHostSize,
+			headers->hostSeparatorHeader,
+			"-----------------", "-----------------", "-----------------",
+			"------");
+}
+
+
+/*
+ * nodestatePrintNodeState prints the node at the given position in the given
+ * nodesArray, using the nodesArray pre-computed sizes for the dynamic columns.
+ */
+void
+nodestatePrintNodeState(NodeAddressHeaders *headers,
+						CurrentNodeState *nodeState)
+{
+	char hostport[BUFSIZE] = { 0 };
+	char composedId[BUFSIZE] = { 0 };
+
+	(void) nodestatePrepareNode(headers,
+								&(nodeState->node),
+								nodeState->groupId,
+								hostport,
+								composedId);
+
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
```

Comment:
The state is now a `%18s` when it used to be a `%17s`, and when refactoring the code I mistakenly kept the `%17s` that applies to the LSN width. Now fixed, and LSN width is computed dynamically too.

I don't think we should compute the state widths dynamically: one of the use cases for `pg_autoctl show state` is when used with a `watch -n 1` command, and the states are going to change. It would be distracting if each output from the watch command was of a different line length. The other fields (name, host:port, lsn) that are computed dynamically will change less often, and I think it's okay to by all dynamic there.

Reply:
It sohuld be `%19s`, there's two characters too little per column in the output I have above. Rest makes sense.

Reply:
Ah yeah. I have now made it internally "dynamic" even though forced to MAX_NODE_STATE_LEN always, and this constant is now edited in the `state.h` header, which is where we add new states. Should make it easier to maintain in the future.